src/ray/raylet/scheduling/policy/node_label_scheduling_policy.cc

```diff
@@ -0,0 +1,189 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "ray/raylet/scheduling/policy/node_label_scheduling_policy.h"
+
+#include "ray/raylet/scheduling/policy/scheduling_context.h"
+
+namespace ray {
+namespace raylet_scheduling_policy {
+
+scheduling::NodeID NodeLabelSchedulingPolicy::Schedule(
+    const ResourceRequest &resource_request, SchedulingOptions options) {
+  RAY_CHECK(options.scheduling_type == SchedulingType::NODE_LABEL);
+  auto context =
+      dynamic_cast<const NodeLabelSchedulingContext *>(options.scheduling_context.get());
+  const auto &scheduling_strategy = context->GetSchedulingStrategy();
+  RAY_CHECK(scheduling_strategy.has_node_label_scheduling_strategy());
+  const auto &node_label_scheduling_strategy =
+      scheduling_strategy.node_label_scheduling_strategy();
+  auto candidate_nodes = SelectAvailableNode(resource_request);
+  if (candidate_nodes.empty()) {
+    return scheduling::NodeID::Nil();
+  }
+  if (node_label_scheduling_strategy.hard().expressions().size() > 0) {
+    SelectNodesByHardExpressions(candidate_nodes, node_label_scheduling_strategy.hard());
+    if (candidate_nodes.empty()) {
+      return scheduling::NodeID::Nil();
+    }
+  }
+
+  const auto &soft_expressions = node_label_scheduling_strategy.soft();
+  if (soft_expressions.expressions().size() > 0) {
+    // soft match
+    auto match_count_map = ScoreNodesBySoftExpressions(candidate_nodes, soft_expressions);
+    auto best_node_id =
+        SelectBestNode(match_count_map, soft_expressions.expressions().size());
+    if (!best_node_id.IsNil()) {
+      return best_node_id;
+    }
+  }
+
+  std::uniform_int_distribution<int> distribution(0, candidate_nodes.size() - 1);
+  int node_index = distribution(gen_);
+  auto it = candidate_nodes.begin();
+  std::advance(it, node_index);
+  return it->first;
+}
+
+scheduling::NodeID NodeLabelSchedulingPolicy::SelectBestNode(
+    const absl::flat_hash_map<scheduling::NodeID, int> &match_count_map,
+    int soft_expressions_size) {
+  std::vector<scheduling::NodeID> all_match_nodes;
+  scheduling::NodeID better_match_node;
+  int better_match_count = 0;
+  for (const auto &[node_id, count] : match_count_map) {
+    if (count == soft_expressions_size) {
+      all_match_nodes.push_back(node_id);
+    } else if (count > better_match_count) {
+      better_match_count = count;
+      better_match_node = node_id;
+    }
+  }
+  if (!all_match_nodes.empty()) {
+    std::uniform_int_distribution<int> distribution(0, all_match_nodes.size() - 1);
+    int node_index = distribution(gen_);
+    return all_match_nodes[node_index];
+  }
+  if (better_match_count > 0) {
+    return better_match_node;
+  }
+  return scheduling::NodeID::Nil();
+}
+
+void NodeLabelSchedulingPolicy::SelectNodesByHardExpressions(
```

Comment:
I think it's cleaner to return a new candidate nodes map instead of modifying the passed-in parameter.

Reply:
Although ` return a new candidate nodes` would make it cleaner, it would repeatedly create and destroy local variables, which can reduce performance. Therefore, I opted to write it this way.

Reply:
If you think it's cleaner as well, then let's do it first. If later on it turns out to be a performance issue, we can change it at that time. Just want to avoid premature optimizations.